### DISCUSSION TITLE
What UI Toolkits Are People Using?

### DISCUSSION THREAD

**Discussion by spiffware** [pelicanmapping/osgearth#2051] [Support Q&A]
I used osgEarth back in the 2.x days, and integrated it seamlessly with Qt applications. Skip ahead to today, and I started building a new Qt application that I planned to use osgEarth in, only to discover that there is no more support for Qt in osgEarth.

What UI framework are people using to develop applications?

I saw another thread where someone had some success using the older osgEarth Qt code with the current 3.2 code base. Is this the best/only way to get osgEarth into Qt?

**Comment by damiandixon** [comment]
I'm using Qt. I used the old integration with a few changes. The project has been on going for about 5yrs so I can't easily change the UI.

For new work I'll probably use Qt but not use the old integration. I will use offscreen rendering for osgearth with separate OpenGL contexts for any new work.

**Reply by spiffware** [reply]
I am developing on Windows 10 using MinGW and have used vcpkg to install osgEarth 3.2 and I'm trying to use Qt 6.x with QtCreator, but I can't get the CMake build environment setup correctly. I have tried a bunch of different things, including appended to the CMAKE_MODULE_PATH, setting the CMAKE_TOOLCHAIN_FILE, etc. I finally got the include path setup for osgEarth, but simply including one of the osgEarth headers causes compile issues with `<thread>`.

What toolchain are you using and on what platform?

Do you happen to have a CMakeLists.txt file that you can share that works with an vcpkg-managed install of osgEarth?

**Comment by jasonbeverage** [comment]
You can definitely use osgEarth with QT, we have lots of customers that are doing it but they are using their own QT integration and not using osgQT which was difficult to support and keep up with.  I'm a huge fan of Dear ImGui and we've started to use that as much as we can for interfaces.  You can check out the osgearth_imgui example to see what we've done there.

**Reply by spiffware** [reply]
Yeah, I've created many Qt apps that use osgEarth, but that was a while ago. I'm not opposed to using a new UI framework, but this particular effort is a very quick prototype, and so I need to use something that I already know, and I'm already very proficient with Qt.

I've basically hit a wall with setting up a working development environment (Qt 6.x, vcpkg install of osgEarth 3.2, and MinGW). I wish I could find a prototype/example that I could start with. I might try switching to a Linux dev environment - I've already run into compatibility issues with some of the MinGW header files (nothing to do with osgEarth).

**Comment by emminizer** [comment]
We use Qt with the [SIMDIS SDK](https://github.com/USNavalResearchLaboratory/simdissdk)  but we have been using an older Qt version (5.9.8) with [osgQt](https://github.com/openscenegraph/osgQt). We have had some issues with newer Qt versions; QOpenGLWidget has not integrated as well as we hoped, so we're still using the older QGLWidget. Because of licensing issues and some other problems we've had in particular with QDockWidget in newer versions, we haven't made the leap to Qt 6.x yet. We also do not currently have vcpkg support.

So, there are some projects out there using it, and it's a bit brittle, but we've been able to make it work. Hope this helps.

**Reply by osgwraiESC** [reply]
upvote what emminizer said, using Qt 5.9.7 with osgQt (version that supports QGLWidget, downloaded a couple years ago) on linux, if you grab the ViewerWidget / ViewerWidget.cpp files out of osgEarth v 2.10 (in the osgEarthQt folder i think) and integrate into your project you can make it work with osgEarth v3.2, see osgearth_qt_simple.cpp from osgEarth v2.10

**Reply by urmjs** [reply]
Hi @emminizer ,
Is it possible to mention the osgQt tag or commit that is currently being distributed as part of the 3rd party library release in https://github.com/USNavalResearchLaboratory/simdissdk/releases/download/simdissdk-1.16/SIMDIS_SDK_3rd_win64_vc-14.2.zip, or if there is a separate github repo of osgQt that is being used with simdis-sdk ?

Thanks.

**Reply by emminizer** [reply]
@urmjs  Of course. We do a detached HEAD at https://github.com/openscenegraph/osgQt/commit/e2a349fd25d14b for our builds.
